### PR TITLE
Get `boost.thread` building

### LINKED ISSUE #2: Missing `srcs` in BUILD file
I am unfamiliar with bazel's inner workings and want to help this project so can you explain why no `srcs` section of `BUILD` files like https://github.com/bazelboost/thread/blob/bazelboost-1.83.0/BUILD.bazel. 

I am asking because I am trying to migrate my build of a large repo from nelhage boost to this one and during linking I am getting a lot of "Undefined symbols for architecture arm64" and am going to try and fix it. 

If you have any pointers of how to address this then please let me know.

### PR DESCRIPTION
**Pull request by andrewkatson**
`boost.thread` builds now. The `test` directory has no `CMakeLists.txt` file so I don't know how to setup the tests so I didn't bother. Also the top level `.txt` file says this should be fine since it has an if clause looking for the subdirectory `.txt` file and does nothing with tests if it does not exist.

### CODE REVIEW

**Comment by zaucy** [comment]
the tests might not have a cmake file, but there's a jamfile. Even just having 1 test running would be really helpful. Doesn't have to be all of them.

**Comment by zaucy** [comment]
I can merge without the test though. I'll probably add it after when I do some testing.

**Comment by andrewkatson** [comment]
Oh Ill add tests. I wasn't sure if I should since there was nothing to go on but I can do it. So hold off on merging.

**Comment by zaucy** [comment]
Thanks! Make sure there's a separate module for the test. Its a great way to test how people will experience the module downstream and useful for separating test dependencies.

**Comment by andrewkatson** [comment]
Gotcha. I have the tests all rounded up but the build is failing because `boost.system` is missing `.ipp` files so I added a PR for that: https://github.com/bazelboost/system/pull/1

**Comment by zaucy** [comment]
merged! should be available in the registry as version `1.83.0.bzl.2` now :+1:

**Comment by zaucy** [comment]
btw even if `boost.system` wasn't released you could test it with `git_override` to get the CI passing.

**Comment by andrewkatson** [comment]
I added tests and it now requires that `boost.chrono` have `srcs` in it. I also excluded various tests that had weird things like missing headers, or deprecated header files, or assigning non const lvalues to string or missing bindings etc... I figured that was out of the scope of this change. Is it fine to push now?

**Comment by zaucy** [comment]
Need a test module so the test dependencies don't get added to the main module. Please add `test/MODULE.bazel`.

File: test/Module.bazel
Comment (zaucy) [inline test/Module.bazel]: test modules don't get published so they don't need a version or compatibility level

File: test/Module.bazel
Comment (zaucy) [inline test/Module.bazel]: `boost.thread` should be added in here as a `local_path_override` https://bazel.build/rules/lib/globals/module#local_path_override

once you have that, most of these dependencies will probably not be required.

File: test/Module.bazel
Comment (zaucy) [inline test/Module.bazel]: not sure if casing matters but this file should be `MODULE.bazel`, not `Module.bazel`

File: test/Module.bazel
Comment (andrewkatson) [reply test/Module.bazel]: Done.

**Comment by zaucy** [comment]
added some github actions that I have on other boost module repo

**Comment by andrewkatson** [comment]
Ok I got it to this point.  Chrono needs to be updated. Do you want me to do that first then this one? 
```
ERROR: /Users/andrewkatson/thread/test/BUILD.bazel:3:8: Linking tests failed: (Exit 1): cc_wrapper.sh failed: error executing CppLink command (from target //:tests) external/bazel_tools~cc_configure_extension~local_config_cc/cc_wrapper.sh @bazel-out/darwin_arm64-fastbuild/bin/tests-2.params

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
Undefined symbols for architecture arm64:
  "boost::chrono::steady_clock::now()", referenced from:
      bool boost::condition_variable::wait_for<long long, boost::ratio<1l, 1000l>, bool (*)()>(boost::unique_lock<boost::mutex>&, boost::chrono::duration<long long, boost::ratio<1l, 1000l>> const&, bool (*)()) in test_11796.o
      bool boost::condition_variable::wait_until<boost::chrono::steady_clock, boost::chrono::duration<long long, boost::ratio<1l, 1000000000l>>, bool (*)()>(boost::unique_lock<boost::mutex>&, boost::chrono::time_point<boost::chrono::steady_clock, boost::chrono::duration<long long, boost::ratio<1l, 1000000000l>>> const&, bool (*)()) in test_11796.o
      bool boost::condition_variable::wait_for<long long, boost::ratio<1l, 1l>, bool (*)()>(boost::unique_lock<boost::mutex>&, boost::chrono::duration<long long, boost::ratio<1l, 1l>> const&, bool (*)()) in test_12293.o
      bool boost::condition_variable::wait_for<int, boost::ratio<3600l, 1l>, bool (*)()>(boost::unique_lock<boost::mutex>&, boost::chrono::duration<int, boost::ratio<3600l, 1l>> const&, bool (*)()) in test_366_1.o
      boost::cv_status boost::condition_variable::wait_for<long long, boost::ratio<1l, 1l>>(boost::unique_lock<boost::mutex>&, boost::chrono::duration<long long, boost::ratio<1l, 1l>> const&) in test_8960.o
      boost::cv_status boost::condition_variable::wait_until<boost::chrono::steady_clock, boost::chrono::duration<long long, boost::ratio<1l, 1000000000l>>>(boost::unique_lock<boost::mutex>&, boost::chrono::time_point<boost::chrono::steady_clock, boost::chrono::duration<long long, boost::ratio<1l, 1000000000l>>> const&) in test_8960.o
      boost::cv_status boost::condition_variable::wait_until<boost::chrono::steady_clock, boost::chrono::duration<long long, boost::ratio<1l, 1000000000l>>>(boost::unique_lock<boost::mutex>&, boost::chrono::time_point<boost::chrono::steady_clock, boost::chrono::duration<long long, boost::ratio<1l, 1000000000l>>> const&) in test_8960.o
      ...
  "boost::chrono::system_clock::now()", referenced from:
      thread() in test_4882.o
      boost::cv_status boost::condition_variable::wait_until<boost::chrono::steady_clock, boost::chrono::duration<long long, boost::ratio<1l, 1000000000l>>>(boost::unique_lock<boost::mutex>&, boost::chrono::time_point<boost::chrono::steady_clock, boost::chrono::duration<long long, boost::ratio<1l, 1000000000l>>> const&) in test_8960.o
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
Target //:tests failed to build
Use --verbose_failures to see the command lines of failed build steps.
INFO: Elapsed time: 30.010s, Critical Path: 3.45s
INFO: 125 processes: 11 internal, 114 darwin-sandbox.
ERROR: Build did NOT complete successfully
//:tests   
```

**Comment by zaucy** [comment]
if boost.thread requires boost.chrono then doing that one first is probably a good idea. If it's optional though, then we could configure it to not use it (or exclude the tests the require it). Either way we should make sure the CI is passing before merging.

**Comment by andrewkatson** [comment]
Okay Ill go do chrono and get back to this later!

File: test/MODULE.bazel
Comment (zaucy) [inline test/MODULE.bazel]: no version required (and not recommended) when doing a path override.

```suggestion
bazel_dep(name = "boost.thread")
```

A lot of the times I put the `bazel_dep` and `local_path_override` side by side to show this, but that doesn't matter.

Comment (andrewkatson) [reply test/MODULE.bazel]: Done

File: test/Module.bazel
Comment (andrewkatson) [reply test/Module.bazel]: Done

File: test/Module.bazel
Comment (andrewkatson) [reply test/Module.bazel]: Done

**Comment by andrewkatson** [comment]
Hey quick question. Each boost test defines its own main method so I cannot glob them into one `cc_test` rule. What is the solution here? Define a `cc_test` per test file? Write my own test main method and remove all the others? Or is there some bazel magic?

**Comment by zaucy** [comment]
Yea defining a `cc_test` per test is pretty common. You can use a "build macro" to make it easier though. Here's an example here: https://github.com/bazelboost/asio/blob/bazelboost-1.83.0/test/BUILD.bazel

multiple `cc_test`s also have better incremental builds :+1:

**Comment by andrewkatson** [comment]
Another question. Some of the tests have no `main` method just `BOOST_AUTO_TEST_CASE` macros. But that throws this kind of error. 

```
ERROR: /Users/andrewkatson/thread/test/BUILD.bazel:91:9: Linking test_condition_notify_one failed: (Exit 1): cc_wrapper.sh failed: error executing CppLink command (from target //:test_condition_notify_one) external/bazel_tools~cc_configure_extension~local_config_cc/cc_wrapper.sh @bazel-out/darwin_arm64-fastbuild/bin/test_condition_notify_one-2.params

Use --sandbox_debug to see verbose messages from the sandbox and retain the sandbox build root for debugging
Undefined symbols for architecture arm64:
  "_main", referenced from:
      <initial-undefines>
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
```

I looked online and found [BOOST_TEST_DYN_LINK](https://www.boost.org/doc/libs/1_84_0/libs/test/doc/html/boost_test/utf_reference/link_references/link_boost_test_dyn_link.html) but that didn't seem to work. Do you know what to do?

**Comment by zaucy** [comment]
@andrewkatson I believe you can include the `boost.test` main with one of the targets found in `@boost.test//:*` such as `@boost.test//:unit_test_main` https://github.com/bazelboost/test/blob/bazelboost-1.83.0/BUILD.bazel#L58-L62

also I updated the deps with the now published new version of bazelboost chrono! thanks for that.

**Review by zaucy — APPROVED** [review]
approving so CI will run automatically

**Comment by andrewkatson** [comment]
Hey so `boost::future` is required by several tests and `boost::thread` but from what I can tell it doesn't build because of this 

```
external/boost.thread~/include/boost/thread/future.hpp:3257:59: error: non-const lvalue reference to type 'basic_string<...>' cannot bind to a temporary of type 'basic_string<...>'
                        this->mark_finished_with_result(f(boost::move(args)...));
                                                          ^~~~~~~~~~~~~~~~~
external/boost.thread~/include/boost/thread/future.hpp:3220:17: note: in instantiation of member function 'boost::detail::task_shared_state<std::string (*)(std::string &), std::string (std::string &)>::do_run' requested here
                task_shared_state(CallableType f_):
                ^
external/boost.thread~/include/boost/thread/future.hpp:3581:33: note: in instantiation of member function 'boost::detail::task_shared_state<std::string (*)(std::string &), std::string (std::string &)>::task_shared_state' requested here
            task = task_ptr(new task_shared_state_type(boost::forward<F>(f)));
                                ^
test_9303.cpp:92:61: note: in instantiation of function template specialization 'boost::packaged_task<std::string (std::string &)>::packaged_task<std::string (&)(std::string &)>' requested here
            boost::packaged_task<std::string(std::string&)> example(string_with_params);
                                                            ^
In file included from test_9303.cpp:15:
In file included from external/boost.thread~/include/boost/thread.hpp:24:
external/boost.thread~/include/boost/thread/future.hpp:3234:58: error: non-const lvalue reference to type 'basic_string<...>' cannot bind to a temporary of type 'basic_string<...>'
                        this->set_value_at_thread_exit(f(boost::move(args)...));
                                                         ^~~~~~~~~~~~~~~~~
external/boost.thread~/include/boost/thread/future.hpp:3220:17: note: in instantiation of member function 'boost::detail::task_shared_state<std::string (*)(std::string &), std::string (std::string &)>::do_apply' requested here
                task_shared_state(CallableType f_):
                ^
external/boost.thread~/include/boost/thread/future.hpp:3581:33: note: in instantiation of member function 'boost::detail::task_shared_state<std::string (*)(std::string &), std::string (std::string &)>::task_shared_state' requested here
            task = task_ptr(new task_shared_state_type(boost::forward<F>(f)));
                                ^
test_9303.cpp:92:61: note: in instantiation of function template specialization 'boost::packaged_task<std::string (std::string &)>::packaged_task<std::string (&)(std::string &)>' requested here
            boost::packaged_task<std::string(std::string&)> example(string_with_params);
                                                            ^
In file included from test_9303.cpp:15:
In file included from external/boost.thread~/include/boost/thread.hpp:24:
external/boost.thread~/include/boost/thread/future.hpp:3781:23: error: non-const lvalue reference to type 'basic_string<...>' cannot bind to a temporary of type 'basic_string<...>'
            task->run(boost::move(args)...);
                      ^~~~~~~~~~~~~~~~~
test_9303.cpp:94:20: note: in instantiation of member function 'boost::packaged_task<std::string (std::string &)>::operator()' requested here
            example(ans);
                   ^
external/boost.thread~/include/boost/thread/future.hpp:2980:56: note: passing argument to parameter 'args' here
            void run(BOOST_THREAD_RV_REF(ArgTypes) ... args)
                                                       ^
external/boost.thread~/include/boost/thread/future.hpp:2995:24: error: non-const lvalue reference to type 'basic_string<...>' cannot bind to a temporary of type 'basic_string<...>'
                do_run(boost::move(args)...);
                       ^~~~~~~~~~~~~~~~~
external/boost.thread~/include/boost/thread/future.hpp:3781:19: note: in instantiation of member function 'boost::detail::task_base_shared_state<std::string (std::string &)>::run' requested here
            task->run(boost::move(args)...);
                  ^
test_9303.cpp:94:20: note: in instantiation of member function 'boost::packaged_task<std::string (std::string &)>::operator()' requested here
            example(ans);
                   ^
external/boost.thread~/include/boost/thread/future.hpp:2979:67: note: passing argument to parameter 'args' here
            virtual void do_run(BOOST_THREAD_RV_REF(ArgTypes) ... args)=0;
```

Any ideas? Do you know how nelhage/boost proper solved this?

**Comment by zaucy** [comment]
I don't think nelhage boost (or the other attempts) added tests for boost.thread. I'm not sure what the problem is.

We could avoid the future related tests and create a gh issue pointing that out if it's the only blocker.

**Comment by andrewkatson** [comment]
Well the thing is we *could* avoid them if we didn't pull in `boost::thread` itself which depends on `boost::future` so imo this is a pretty hard blocker...

**Comment by zaucy** [comment]
the errors only would occur on use. i.e. `test_9303.cpp:94:20` in the callstack you posted above.

**Comment by andrewkatson** [comment]
Yeah but `boost::thread` is used in every other test and it depends on `boost::future`.  Unless I am missing something?

**Comment by zaucy** [comment]
does every test use `boost::future` though? They're template functions, they don't get instantiated until use. So if there's a problem it depends on the callsite, not the declaration.

**Comment by zaucy** [comment]
I pushed a few tweaks which will get some tests to pass (at least on windows) so you can see what I mean :+1:

**Comment by zaucy** [comment]
See. A bunch of tests can pass even though they rely on `boost.thread`. When you're dealing with templates the errors will only happen on-use. So if there's an issue with `boost::future` in regards to `boost.thread` we can avoid it for now by not including those particular tests and make a gh issue about it.

https://github.com/bazelboost/thread/actions/runs/8609220542/job/23592812113?pr=3

```
//:test_11818                                                   FAILED TO BUILD
//:test_13561                                                   FAILED TO BUILD
//:test_2741                                                    FAILED TO BUILD
//:test_3837                                                    FAILED TO BUILD
//:test_8557                                                    FAILED TO BUILD
(Skipping other failed to build tests)
//:test_10340                                                            PASSED in 0.0s
//:test_10964                                                            PASSED in 0.1s
//:test_11053                                                            PASSED in 0.1s
//:test_11256                                                            PASSED in 0.1s
//:test_11266                                                            PASSED in 0.0s
//:test_11499                                                            PASSED in 0.1s
//:test_11611                                                            PASSED in 0.1s
//:test_11796                                                            PASSED in 0.2s
//:test_12293                                                            PASSED in 0.0s
//:test_12949                                                            PASSED in 0.0s
//:test_13480b                                                           PASSED in 0.1s
//:test_2309                                                             PASSED in 0.0s
//:test_2501                                                             PASSED in 0.1s
//:test_3628                                                             PASSED in 5.0s
//:test_366_1                                                            PASSED in 0.0s
//:test_366_2                                                            PASSED in 0.1s
//:test_366_3                                                            PASSED in 0.0s
//:test_366_4                                                            PASSED in 1.0s
//:test_4521                                                             PASSED in 0.0s
//:test_4648                                                             PASSED in 0.1s
//:test_4882                                                             PASSED in 0.7s
//:test_[535](https://github.com/bazelboost/thread/actions/runs/8609220542/job/23592812113?pr=3#step:3:536)1                                                             PASSED in 0.1s
//:test_5542_1                                                           PASSED in 3.1s
//:test_5542_2                                                           PASSED in 0.0s
//:test_5[542](https://github.com/bazelboost/thread/actions/runs/8609220542/job/23592812113?pr=3#step:3:543)_3                                                           PASSED in 3.0s
//:test_5891                                                             PASSED in 2.1s
//:test_6130                                                             PASSED in 0.0s
//:test_6170                                                             PASSED in 0.0s
//:test_6174                                                             PASSED in 0.0s
//:test_7160                                                             PASSED in 0.0s
//:test_7328                                                             PASSED in 0.0s
//:test_7571                                                             PASSED in 0.1s
//:test_7665                                                             PASSED in 30.6s
//:test_7666                                                             PASSED in 5.1s
//:test_7720                                                             PASSED in 0.0s
//:test_7755                                                             PASSED in 1.6s
//:test_8508                                                             PASSED in 0.0s
//:test_8586                                                             PASSED in 0.0s
//:test_8596                                                             PASSED in 0.0s
//:test_8600                                                             PASSED in 0.0s
//:test_8674                                                             PASSED in 0.0s
//:test_8943                                                             PASSED in 0.1s
//:test_8960                                                             PASSED in 1.1s
//:test_9079_a                                                           PASSED in 0.0s
//:test_9319                                                             PASSED in 0.0s
//:test_9711                                                             PASSED in 0.0s
//:test_9856                                                             PASSED in 10.6s
//:test_barrier                                                          PASSED in 0.0s
//:test_barrier_size_fct                                                 PASSED in 0.0s
//:test_barrier_void_fct                                                 PASSED in 0.0s
//:test_completion_latch                                                 PASSED in 0.0s
//:test_latch                                                            PASSED in 0.1s
//:test_ml                                                               PASSED in 0.0s
//:test_ml2                                                              PASSED in 0.0s
//:test_scheduled_tp                                                     PASSED in 10.1s
//:test_scheduling_adaptor                                               PASSED in 10.1s
//:test_thread_mf                                                        PASSED in 0.1s
//:test_8455                                                            TIMEOUT in 300.0s
  C:/users/runneradmin/_bazel_runneradmin/ekd7e7of/execroot/_main/bazel-out/x64_windows-fastbuild/testlogs/test_8455/test.log
//:test_10963                                                            FAILED in 1.1s
  C:/users/runneradmin/_bazel_runneradmin/ekd7e7of/execroot/_main/bazel-out/x64_windows-fastbuild/testlogs/test_10963/test.log
//:test_[550](https://github.com/bazelboost/thread/actions/runs/8609220542/job/23592812113?pr=3#step:3:551)2                                                             FAILED in 0.6s
  C:/users/runneradmin/_bazel_runneradmin/ekd7e7of/execroot/_main/bazel-out/x64_windows-fastbuild/testlogs/test_5502/test.log

Executed 60 out of 93 tests: 57 tests pass, 33 fail to build and 3 fail locally.
```

**Comment by andrewkatson** [comment]
Oh I see what you mean now! Ill skip the failing tests and document which ones fail.

**Comment by andrewkatson** [comment]
I have two tests that still fail but build. Should I try fixing them? We haven't changed anything in Boost itself so it is unclear why it doesn't work.

**Comment by zaucy** [comment]
yeah if a few tests are having issue we can just document it in an issue or something and move forward

often the reason for failure on boost tests is one of these:
a) the test requires a specific configuration that we don't support with bazel yet
b) the test requires some files at runtime which can sometimes be difficult to give to the test with bazel (since we can't use runfiles)

**Comment by andrewkatson** [comment]
Okay awesome Ill just document it.

File: include/boost/thread/pthread/pthread_helpers.hpp
Comment (zaucy) [inline include/boost/thread/pthread/pthread_helpers.hpp]: hard rule in the `bazelboost` repositories is no non-bazel file changes.

**Review by zaucy — CHANGES_REQUESTED** [review]

File: include/boost/thread/pthread/pthread_helpers.hpp
Comment (andrewkatson) [reply include/boost/thread/pthread/pthread_helpers.hpp]: Yeah that is what I figured but wasn't sure. I will remove it. My feeing is we will have to remove a fair few tests though. :(

**Comment by andrewkatson** [comment]
Okay so sad news. I tried and did `--keep_going` and every single test fails to build if you exclude the *one* file (`pthread_helpers.hpp`) from the build of `@boost.thread`. So until that file gets fixed (with the assert import and other changes probably) then we cannot get any tests to run on Posix. Unless there is a way in bazel to get this to work? I was thinking maybe generate a file or progrmatically modify `pthread_helpers.hpp"

**Comment by zaucy** [comment]
wow yeah that is a real bummer. I'm not sure how they're getting that to work 🤔

**Comment by zaucy** [comment]
@andrewkatson it might be this: https://github.com/bazelboost/thread/blob/f339fa274e5b559ccfa91385dc21c37957ebcfd6/test/Jamfile.v2#L279-L280

defining `BOOST_THREAD_TEST_HEADER` with something might help.

**Comment by andrewkatson** [comment]
> @andrewkatson it might be this:
> 
> https://github.com/bazelboost/thread/blob/f339fa274e5b559ccfa91385dc21c37957ebcfd6/test/Jamfile.v2#L279-L280
> 
> defining `BOOST_THREAD_TEST_HEADER` with something might help.

I have been trying for the last 2 hours to in place modify a file with bazel genrules and/or custom rules and as far as I can tell it cannot be done. There is a hard limit on making a src == to an output and there is a cycle created if you do something like move a fileA > fileB then modify fileB and move it back to where fileA was with two or more genrules. So even if I generate a new boost header file I would need to include it in existing files afaik I cannot do that. The only other option I see is to modify a series of files so that all the includes are changed. So instead of `pthread_helpers.hpp` you would have `helpers.hpp` and you would need to replace that across files and then replace those files everywhere and it gets really brittle. LMK if you have any other ideas because I am scratching my head. I get why bazel doesn't allow the in place modification of files with genrules but it is an annoying limitation.

**Comment by andrewkatson** [comment]
Okay I managed to get bazel to answer and they said to ask boost to modify the files if we want things to work. How realistic is that?

**Comment by andrewkatson** [comment]
Outside of adding `#include <boost/assert.hpp>` to `pthread_helpers.hpp` everything else will work now with my latest commit. and we get to keep ~80 tests.

**Comment by andrewkatson** [comment]
@zaucy I tried defining `BOOST_THREAD_TEST_HEADER` and that did not fix things. What should we do?

**Comment by zaucy** [comment]
I'll be able to look into this in the next few days. Will post back then!

**Comment by andrewkatson** [comment]
No problem! :) I get so eager

**Comment by andrewkatson** [comment]
Hey any update on what we do with this?

**Comment by zaucy** [comment]
Ill take a look today! Thanks for the reminder.

**Comment by zaucy** [comment]
I think we have no choice but to modify the the source. I added a PR upstream to make the change more official though https://github.com/boostorg/thread/pull/401

**Comment by andrewkatson** [comment]
Thank you for making the change to boost! Yeah its an odd situation but at least its only an added header not major structural changes.

**Comment by zaucy** [comment]
@andrewkatson should be available as version `1.83.0.bzl.2` in the registry! Thanks for all your hard work and patience.

**Comment by andrewkatson** [comment]
AWESOME! thank you!!

**Pull request merged**
